### ISSUE TITLE
v13

### ISSUE THREAD

**Issue by JamesHenry** [angular-eslint/angular-eslint#715]
## Update on v13 (as of Nov 3, 2021)

The first prerelease has been published, anybody wanting to migrate to v13 can run:

```sh
npx ng update @angular-eslint/schematics --next
```

OR (for new workspaces)

```sh
npx ng add @angular-eslint/schematics@next
```

**After** they have already upgraded to Angular v13.

Please note: more breaking changes are likely at this stage before we declare v13 of angular-eslint stable, so please be aware of that when deciding whether or not to upgrade.

We will cut the stable release within the next week or so.

The PR is here if you wish to subscribe for updates: https://github.com/angular-eslint/angular-eslint/pull/780

---

This is the summary issue for work which will be included in v13 of the `@angular-eslint/*` packages.

The primary goals are:

- [x] Support angular 13
- [x] Support typescript-eslint 5
- [x] Support eslint 8
- [x] Support TypeScript 4.4

We will look to include other minor work which require breaking changes alongside this, but those are the priority items.

We will create a v13 branch and publish prereleases from it alongside patches to v12 from the master branch until `angular-eslint` v13 is deemed stable.

---

## ~Update on supporting Angular 13 (as of Oct 25, 2021)~

~Between me and the Angular Team we have decided to start bundling the `@angular/compiler` within the `@angular-eslint` repo as a new package (which should never need to care about) called `@angular-eslint/bundled-angular-compiler`.~

~As reported below, the original reason to consider this was the fact that the Angular Team has stopped publishing bundles which can be synchronously `require`d in node as of v13, but actually in discussing it further with the team there is another important reason to start bundling ourselves - they are planning on making a larger number of breaking changes to the internals of the compiler from v13 onwards, and the APIs we depend on are not officially part of the public API and so are not subject to the normal major version schedule.~

~Thanks to Alan from the Angular Tooling Team, we have already landed the new package and the way is now clear to focus on the remaining breaking changes related to Angular 13.~

## ~Notes on supporting Angular 13 (as of Oct 17, 2021)~

~Currently in the v13 prereleases the Angular Team has dropped all non-ESM artifacts from the packages. This is a huge problem for us because ESLint will not allow any of our tooling to be async and we therefore cannot resolve the runtime elements of `@angular/compiler` we need from `node_modules` in a way that is compatible with Node 12, which is still not EOL and is rightly supported by ESLint.~

~As things stand we would need to bundle an exact patch version of the `@angular/compiler` ourselves and keep this up to date.~

~I have written to the Angular Team on this point.~

**Comment by JounQin** [comment]
Maybe https://github.com/rx-ts/synckit can help to `import()` in worker.

**Comment by JounQin** [comment]
Why a major release is required for eslint 8 support? I try to lock `@typescript-eslint/experimental-utils@5.0.0`, it seems just working, is there any known issue?

**Comment by byrondover** [comment]
I'm using the same workaround @JounQin suggests for now, pinning `@typescript-eslint/experimental-utils` in my project's `package.json`:

```json
  "resolutions": {
    "@typescript-eslint/experimental-utils": "5.0.0"
  }
```

Works natively with `yarn`, requires [npm-force-resolutions](https://www.npmjs.com/package/npm-force-resolutions) with `npm`.

---

I can see why this would be a good opportunity for a major `@angular-eslint` release, although given the note about ESM artifact support in Angular 13, I worry it may take a prohibitively long time. That said, with a decent workaround in hand, I'm personally not too bummed about waiting.

Appreciate all that you do for the community @JamesHenry! 🙇 

Good luck in your discussions with the Angular team, let us know if there are any issues over on `angular/angular` that we can go 👍 vote on.

**Comment by JamesHenry** [comment]
https://github.com/angular-eslint/angular-eslint/pull/780 is nearly ready and I will begin publishing prereleases from that PR later today/tomorrow.

In the meantime I’ve been told by a user that version 12.6.0 functions with Angular 13 (thanks to the fact that we now bundle the compiler I imagine) so hopefully nobody is blocked if they are keen to jump on Angular 13 as soon as it’s released today.

**Comment by JamesHenry** [comment]
The first prerelease has been published, anybody wanting to migrate to v13 can run:

```sh
npx ng update @angular-eslint/schematics --next
```

**After** they have already upgraded to Angular v13.

Please note: more breaking changes are likely at this stage before we declare v13 of angular-eslint stable, so please be aware of that when deciding whether or not to upgrade.

We will cut the stable release within the next week or so.

**Comment by michaelfaith** [comment]
Thanks for being on top of the update.  Appreciate the work on this.

**Comment by gagle** [comment]
Still happening with 13.0.0-alpha.0

**Comment by michaelfaith** [comment]
> Still happening with 13.0.0-alpha.0
> 
What is?

**Comment by pavankjadda** [comment]
@JamesHenry When I run the command `npx ng add @angular-eslint/schematics --next` in a tslint project. I get the error message 
```
An unhandled exception occurred: NOT SUPPORTED: keyword "id", use "$id" for schema ID
See "/private/var/folders/1v/vz08n4553r56bprqn9gcmqtw7ggw12/T/ng-O9SAxp/angular-errors.log" for further details.
```

**Comment by umairhm** [comment]
> @JamesHenry When I run the command `npx ng add @angular-eslint/schematics --next` in a tslint project. I get the error message
> 
> ```
> An unhandled exception occurred: NOT SUPPORTED: keyword "id", use "$id" for schema ID
> See "/private/var/folders/1v/vz08n4553r56bprqn9gcmqtw7ggw12/T/ng-O9SAxp/angular-errors.log" for further details.
> ```

What's your node version? I think I had to move from Node 16 to Node 14 to get rid of this error.

**Comment by prabh-62** [comment]
`yarn ng add @angular-eslint/schematics@next` worked for me

**Comment by erksmaas** [comment]
After installing npx ng update @angular-eslint/schematics --next on Angular 13, we are seeing:

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Failed to load plugin 'eslint-plugin-import' declared in '.eslintrc.json#overrides[0] » plugin:@angular-eslint/ng-cli-compat': Package subpath './lib/util/glob-util' is not defined by "exports" in C:\finch\customer\node_modules\eslint\package.json

**Comment by theodorejb** [comment]
I'm getting the same error as @erksmaas.

**Comment by jtth** [comment]
> After installing npx ng update @angular-eslint/schematics --next on Angular 13, we are seeing:
> 
> Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Failed to load plugin 'eslint-plugin-import' declared in '.eslintrc.json#overrides[0] » plugin:@angular-eslint/ng-cli-compat': Package subpath './lib/util/glob-util' is not defined by "exports" in C:...\node_modules\eslint\package.json

I'm just trying things out, but removing the following from the eslint settings removed the exception for me:

```
//        "plugin:@angular-eslint/ng-cli-compat",
//        "plugin:@angular-eslint/ng-cli-compat--formatting-add-on",
```

**Comment by erksmaas** [comment]
Same here

**Comment by JamesHenry** [comment]
Thanks for your patience everyone, v13.0.0 of angular-eslint has been released, if you encounter any issues with the stable version please kindly open new issues and follow the provided issue templates.

**Comment by michaelfaith** [comment]
Awesome. Thanks for all your hard work!

**Comment by erksmaas** [comment]
Thanks.  I'm still seeing the ng-cli-compat errors after switching to v13.0.1 FYI

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Failed to load plugin 'eslint-plugin-import' declared in '.eslintrc.json#overrides[0] » plugin:@angular-eslint/ng-cli-compat': Package subpath './lib/util/glob-util' is not defined by "exports" in C:...\node_modules\eslint\package.json

**Comment by JamesHenry** [comment]
@erksmaas If you bumped to 13.0.1 manually that is expected because it is the version of `eslint-plugin-import` is the issue.

For users running `npx ng update @angular-eslint/schematics` any time since 13.0.1 was released yesterday, the schematic will handle updating `eslint-plugin-import` to version `2.25.2` automatically.

For anyone that ran `npx ng update @angular-eslint/schematics` before 13.0.1 came out, or anyone who has manually bumped versions, you will need to update `eslint-plugin-import` to version `2.25.2` or above yourselves.

**Comment by gagle** [comment]
I still get this error with latest versions of everything :(

````
Failed to load plugin '@nrwl/nx' declared in 'apps/XXX/.eslintrc.json » ../../.eslintrc.json#overrides[0]': Class extends value undefined is not a constructor or null
````

**Comment by nzacca** [comment]
@gagle I believe we're waiting on these commits to be released by the nx team before this error can be fixed: https://github.com/nrwl/nx/compare/13.2.2...master

In particular, this PR contains support for eslint 8: https://github.com/nrwl/nx/pull/7800

**Comment by indraraj26** [comment]
> @erksmaas If you bumped to 13.0.1 manually that is expected because it is the version of `eslint-plugin-import` that is the issue.
> 
> For users running `npx ng update @angular-eslint/schematics` any time since 13.0.1 was released yesterday, the schematic will handle updating `eslint-plugin-import` to version `2.25.2` automatically.
> 
> For anyone that ran `npx ng update @angular-eslint/schematics` before 13.0.1 came out, or anyone who has manually bumped versions, you will need to update `eslint-plugin-import` to version `2.25.2` or above yourselves.
> 
> ESLint 8 had breaking changes for plugin maintainers, and that is the version of `eslint-plugin-import` which supports ESLint 8.

Getting this error
```
npx ng update @angular-eslint/schematics  
The installed local Angular CLI version is older than the latest stable version.
Installing a temporary version to perform the update.
✔ Package successfully installed.
Using package manager: 'npm'
Collecting installed dependencies...
Found 28 dependencies.
Package '@angular-eslint/schematics' is not a dependency.

```

ng version
```
ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 12.1.0
Node: 14.17.1
Package Manager: npm 6.14.13
OS: win32 x64

Angular: 12.1.5
... common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1201.4
@angular-devkit/build-angular   12.1.4
@angular-devkit/core            12.1.4
@angular-devkit/schematics      12.1.4
@angular/cli                    12.1.4
@angular/language-service       12.0.5
@schematics/angular             12.1.4
rxjs                            6.6.7 (cli-only)
typescript                      <error>

```

**Comment by michaelfaith** [comment]
Maybe it's a silly question, but do you have `@angular-eslint/schematics` installed in the package where you ran the update command?

**Comment by indraraj26** [comment]
Yes i do have and i can see it in node-modules. I am running this command in root directory of project where package.json and package-lock.json lives. Also i can see `@angular-eslint/schematics` entry in under devDependency of package.json

**Comment by dertuerke** [comment]
Hi, i get this warning on a fresh angular-project if i run `ng lint`
```
WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <4.5.0

YOUR TYPESCRIPT VERSION: 4.5.4

Please only submit bug reports when using the officially supported version.
```

**Comment by michaelfaith** [comment]
@dertuerke may want to open a new issue.  I believe Angular started shipping with Typescript 4.5 in 13.1.  In 13.0 it was still 4.4

**Comment by vanessawa** [comment]
I have the same Problem while updating from v13 to v14. 
But how should I run this "After upgrading to Angular v14"? The problem is when running the update command.
![image](https://user-images.githubusercontent.com/81361169/223062508-3bfab3d5-b674-44ed-8fba-5aea8e002bf1.png)
![image](https://user-images.githubusercontent.com/81361169/223063281-be188264-b4b5-48b9-bb0e-8a91643eda68.png)

**Comment by dertuerke** [comment]
@vanessawa you need to force update first `angular/core` and `angular/cli`, then you can update `angular/eslint`

**Comment by vanessawa** [comment]
@dertuerke you mean "ng update @angular/core@14 @angular/cli@14 --force"?

**Comment by dertuerke** [comment]
@vanessawa yes! `ng update angular/cli@14 angular/core@14 --force`

**Comment by vanessawa** [comment]
@dertuerke This did not work:
![image](https://user-images.githubusercontent.com/81361169/223077818-d86b7c18-cba8-4868-b5c1-621434399912.png)
![image](https://user-images.githubusercontent.com/81361169/223078036-bb8d8558-b18f-4a43-8578-846750ed4ac4.png)
